chainstore/chain_store.go

```diff
@@ -0,0 +1,202 @@
+		blkBytes = blk.Bytes()
+	)
+	heightBytes := binary.BigEndian.AppendUint64(nil, height)
+	err := errors.Join(
+		batch.Put([]byte{lastAcceptedByte}, heightBytes),
+		batch.Put(PrefixBlockIDHeightKey(blkID), heightBytes),
+		batch.Put(PrefixBlockHeightIDKey(blk.Height()), blkID[:]),
+		batch.Put(PrefixBlockKey(height), blkBytes),
+	)
+	if err != nil {
+		return err
+	}
+	expiryHeight := height - c.config.AcceptedBlockWindow
+	var expired bool
+	if expiryHeight > 0 && expiryHeight < height { // ensure we don't free genesis
+		if err := batch.Delete(PrefixBlockKey(expiryHeight)); err != nil {
+			return err
+		}
+		blkID, err := c.db.Get(PrefixBlockHeightIDKey(expiryHeight))
+		if err != nil {
+			return fmt.Errorf("unable to fetch blockID at height %d: %w", expiryHeight, err)
+		}
+		if err := batch.Delete(PrefixBlockIDHeightKey(ids.ID(blkID))); err != nil {
+			return err
+		}
+		if err := batch.Delete(PrefixBlockHeightIDKey(expiryHeight)); err != nil {
+			return err
+		}
+		expired = true
+		c.metrics.deletedBlocks.Inc()
+	}
+	//nolint:gosec
+	if expired && rand.Intn(c.config.BlockCompactionAverageFrequency) == 0 {
+		go func() {
+			start := time.Now()
+			if err := c.db.Compact([]byte{blockPrefix}, PrefixBlockKey(expiryHeight)); err != nil {
+				c.log.Error("failed to compact block store", zap.Error(err))
+				return
+			}
+			c.log.Info("compacted disk blocks", zap.Uint64("end", expiryHeight), zap.Duration("t", time.Since(start)))
+		}()
+	}
+
+	return batch.Write()
+}
+
+func (c *ChainStore[T]) GetBlock(ctx context.Context, blkID ids.ID) (T, error) {
+	var emptyT T
+	height, err := c.GetBlockIDHeight(ctx, blkID)
+	if err != nil {
+		return emptyT, err
+	}
+	return c.GetBlockByHeight(ctx, height)
+}
+
+func (c *ChainStore[T]) GetBlockIDAtHeight(_ context.Context, blkHeight uint64) (ids.ID, error) {
+	blkIDBytes, err := c.db.Get(PrefixBlockHeightIDKey(blkHeight))
+	if err != nil {
+		return ids.Empty, err
+	}
+	return ids.ID(blkIDBytes), nil
+}
+
+func (c *ChainStore[T]) GetBlockIDHeight(_ context.Context, blkID ids.ID) (uint64, error) {
+	blkHeightBytes, err := c.db.Get(PrefixBlockIDHeightKey(blkID))
+	if err != nil {
+		return 0, err
+	}
+	return database.ParseUInt64(blkHeightBytes)
+}
+
+func (c *ChainStore[T]) GetBlockByHeight(ctx context.Context, blkHeight uint64) (T, error) {
+	var emptyT T
+	blkBytes, err := c.db.Get(PrefixBlockKey(blkHeight))
+	if err != nil {
+		return emptyT, err
+	}
+	return c.parser.ParseBlock(ctx, blkBytes)
+}
+
+func PrefixBlockKey(height uint64) []byte {
+	k := make([]byte, 1+consts.Uint64Len)
+	k[0] = blockPrefix
+	binary.BigEndian.PutUint64(k[1:], height)
+	return k
+}
+
+func PrefixBlockIDHeightKey(id ids.ID) []byte {
+	k := make([]byte, 1+ids.IDLen)
+	k[0] = blockIDHeightPrefix
+	copy(k[1:], id[:])
+	return k
+}
+
+func PrefixBlockHeightIDKey(height uint64) []byte {
+	k := make([]byte, 1+consts.Uint64Len)
+	k[0] = blockHeightIDPrefix
+	binary.BigEndian.PutUint64(k[1:], height)
+	return k
+}
```

Comment:
Should these be un-exported?

Reply:
Yup, these are moved not changed, but I can un-export them.